source/cppfs/source/windows/LocalFileWatcher.cpp

```diff
@@ -0,0 +1,196 @@
+
+#include <cppfs/windows/LocalFileWatcher.h>
+
+#include <algorithm>
+
+#include <Windows.h>
+
+#include <cppfs/windows/LocalFileSystem.h>
+
+
+namespace
+{
+    struct LocalWatcher
+    {
+        OVERLAPPED ovl;
+        std::shared_ptr<VOID> event;
+        DWORD buffer[16384];
+    };
+
+    // NB! Using a std::mutex resulted in spurious deadlocks using VS2017, whilst
+    // using CRITICAL_SECTION works fine.
+    struct ScopedCriticalSection
+    {
+        LPCRITICAL_SECTION m_pcs;
+        ScopedCriticalSection(LPCRITICAL_SECTION pcs) : m_pcs(pcs) {
+            ::EnterCriticalSection(m_pcs);
+        }
+        ~ScopedCriticalSection() {
+            ::LeaveCriticalSection(m_pcs);
+        }
+    };
+}
+
+
+namespace cppfs
+{
+
+
+LocalFileWatcher::LocalFileWatcher(FileWatcher * fileWatcher, std::shared_ptr<LocalFileSystem> fs)
+: AbstractFileWatcherBackend(fileWatcher)
+, m_fs(fs)
+, m_waitStopEvent(::CreateEvent(NULL, TRUE, FALSE, NULL), ::CloseHandle)
+, m_watchersCS(new CRITICAL_SECTION())
+{
+    ::InitializeCriticalSection((LPCRITICAL_SECTION)m_watchersCS.get());
+}
+
+LocalFileWatcher::~LocalFileWatcher()
+{
+    ::DeleteCriticalSection((LPCRITICAL_SECTION)m_watchersCS.get());
+    m_watchers.clear();
+}
+
+AbstractFileSystem * LocalFileWatcher::fs() const
+{
+    return static_cast<AbstractFileSystem *>(m_fs.get());
+}
+
+void LocalFileWatcher::add(FileHandle & fh, unsigned int events, RecursiveMode recursive)
+{
+    HANDLE hDir = ::CreateFileA(fh.path().c_str(),       // pointer to the directory name
```

Comment:
What happens if fh points to a file? If I read it correctly, you assume that fh points to a directory. fh.path() gives back the path to the file system object (file or directory). If you want the path to the parent directory, you must use fh.directoryPath(), but then you should also make sure that fh is not a directory itself. Maybe we have to make clear our semantics here (what shall happen if I add a directory handle to a file watcher?)

Reply:
There is no provision for listening to a specific file, I assume from msdn docs, so making sure that path is a directory is appropriate.